src/support/hazard.c

```diff
@@ -101,7 +112,8 @@ __wt_hazard_set(WT_SESSION_IMPL *session, WT_REF *ref, bool *busyp
 		 * found this page using the tree's key space, whatever page we
 		 * find here is the page for us to use.)
 		 */
-		if (ref->page == hp->page && ref->state == WT_REF_MEM) {
+		if (ref->page == page && ref->state == WT_REF_MEM) {
+			WT_ASSERT(session, hp->page != NULL);
```

Comment:
On further thought, we may want:

```
if (ref->state == WT_REF_MEM) {
        WT_READ_BARRIER();
        if (ref->page == page) {
                ...
        }
}
```

i.e., what we need to be sure about is that `ref->page` is correct *after* we check `ref->state`.  Let me know what you think: I don't have enough runs yet to be confident that what's in the branch is sufficient.